[rpc] Avoid wireDeserializer overreading buffers by 1 byte

jjlilley opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#36976 [rpc] Avoid wireDeserializer overreading buffers by 1 byte**

The bounds check and the read were swapped in two places - I noticed
ASAN complaining in an unrelated change on an erroneous buffer.

Adding a couple simple test cases.

Differential Revision: [D21148936](https://our.internmc.facebook.com/intern/diff/D21148936/)

Commit:
[rpc] Avoid wireDeserializer overreading buffers by 1 byte

The bounds check and the read were swapped in two places - I noticed
ASAN complaining in an unrelated change on an erroneous buffer.

Adding a couple simple test cases.

Differential Revision: [D21148936](https://our.internmc.facebook.com/intern/diff/D21148936/)

[ghstack-poisoned]

mrshenli commented on test/cpp/rpc/test_wire_serialization.cpp:
```diff
@@ -68,6 +68,13 @@ TEST(WireSerialize, CloneSparseTensors) {
 }
 
+TEST(WireSerialize, Errors) {
+  EXPECT_THROW((void)torch::distributed::rpc::wireDeserialize("", 0), std::runtime_error);
+  EXPECT_THROW((void)torch::distributed::rpc::wireDeserialize(" ", 1), std::runtime_error);
```
Shall we also check the error message here? "failed parse" vs "failed bounds"?

jjlilley replied:
sure, just uploaded a version with this check

mrshenli commented on test/cpp/rpc/test_wire_serialization.cpp:
```diff
@@ -68,6 +68,13 @@ TEST(WireSerialize, CloneSparseTensors) {
+  EXPECT_THROW((void)torch::distributed::rpc::wireDeserialize("", 0), std::runtime_error);
+  EXPECT_THROW((void)torch::distributed::rpc::wireDeserialize(" ", 1), std::runtime_error);
+  auto serialized = torch::distributed::rpc::wireSerialize({}, {torch::randn({5, 5})});
+  EXPECT_THROW((void)torch::distributed::rpc::wireDeserialize(serialized.data(), serialized.size() / 2), std::runtime_error);
```
this should throw "failed parse" as well?

jjlilley replied:
This one is "failed bounds" actually - but thought that we should cover this somewhere

mrshenli reviewed (approved):
Thanks for the fix!

Commit:
Update on "[rpc] Avoid wireDeserializer overreading buffers by 1 byte"


The bounds check and the read were swapped in two places - I noticed
ASAN complaining in an unrelated change on an erroneous buffer.

Adding a couple simple test cases.

Differential Revision: [D21148936](https://our.internmc.facebook.com/intern/diff/D21148936/)

[ghstack-poisoned]

jjlilley commented:
Thanks for looking, just uploaded a version with the extra unittest check
